aws/resource_aws_network_interface_sg_attachment_test.go

```diff
@@ -0,0 +1,238 @@
+		}
+		interfaceID := s.Modules[0].Resources["aws_instance.instance"].Primary.Attributes[ifAttr]
+		sgID := s.Modules[0].Resources["aws_security_group.sg"].Primary.ID
+
+		iface, err := fetchNetworkInterface(conn, interfaceID)
+		if err != nil {
+			return err
+		}
+		actual := sgExistsInENI(sgID, iface)
+		if expected != actual {
+			return fmt.Errorf("expected existence of security group in ENI to be %t, got %t", expected, actual)
+		}
+		return nil
+	}
+}
+
+func testAccAwsNetworkInterfaceSGAttachmentConfigViaInstance(attachmentEnabled bool) string {
+	return fmt.Sprintf(`
+variable "sg_attachment_enabled" {
+  type    = "string"
+  default = "%t"
+}
+
+data "aws_ami" "ami" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["amzn-ami-hvm-*"]
+  }
+
+  owners = ["amazon"]
+}
+
+resource "aws_instance" "instance" {
+  instance_type = "t2.micro"
+  ami           = "${data.aws_ami.ami.id}"
+
+  tags = {
+    "type" = "terraform-test-instance"
+  }
+}
+
+resource "aws_security_group" "sg" {
+  tags = {
+    "type" = "terraform-test-security-group"
+  }
+}
+
+resource "aws_network_interface_sg_attachment" "sg_attachment" {
+  count                = "${var.sg_attachment_enabled == "true" ? 1 : 0}"
+  security_group_id    = "${aws_security_group.sg.id}"
+  network_interface_id = "${aws_instance.instance.primary_network_interface_id}"
+}
+`, attachmentEnabled)
+}
+
+func testAccAwsNetworkInterfaceSGAttachmentConfigViaDataSource(attachmentEnabled bool) string {
+	return fmt.Sprintf(`
+variable "sg_attachment_enabled" {
+  type    = "string"
+  default = "%t"
+}
+
+data "aws_ami" "ami" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["amzn-ami-hvm-*"]
+  }
+
+  owners = ["amazon"]
+}
+
+resource "aws_instance" "instance" {
+  instance_type = "t2.micro"
+  ami           = "${data.aws_ami.ami.id}"
+
+  tags = {
+    "type" = "terraform-test-instance"
+  }
+}
+
+data "aws_instance" "external_instance" {
+  instance_id = "${aws_instance.instance.id}"
+}
+
+resource "aws_security_group" "sg" {
+  tags = {
+    "type" = "terraform-test-security-group"
+  }
+}
+
+resource "aws_network_interface_sg_attachment" "sg_attachment" {
+  count                = "${var.sg_attachment_enabled == "true" ? 1 : 0}"
+  security_group_id    = "${aws_security_group.sg.id}"
+  network_interface_id = "${data.aws_instance.external_instance.network_interface_id}"
+}
+`, attachmentEnabled)
+}
+
+func TestAccAwsNetworkInterfaceSGAttachmentRaceCheck(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAwsNetworkInterfaceSGAttachmentRaceCheckConfig("step1"),
+				Check:  checkSecurityGroupAttachmentRace("step1"),
+			},
+			resource.TestStep{
+				Config: testAccAwsNetworkInterfaceSGAttachmentRaceCheckConfig("step2"),
+				Check:  checkSecurityGroupAttachmentRace("step2"),
+			},
```

Comment:
Step 2 should basically destroy all `aws_network_interface_sg_attachment_step1.*` instances and create `aws_network_interface_sg_attachment_step2.*`. This *should* have the effect of sufficient chaos as 4 detachment and 4 attachment operations are going on at the same time in step 2 (4 resources being destroyed and 4 being created).

I think what might need work is `aws_security_group.sg_step1|step2` - I don't know if maintaining these separately is getting the desired above effect and maybe a central group of SGs should be created instead.